File: src/targeting/ric_frame.cpp
#include "ric_frame.hpp"
#include <cmath>

namespace sim {

RICFrame RICFrameUtils::compute_frame(const StateVector& state) {
    RICFrame ric;

    // R = position / |position|
    double r_mag = state.position.norm();
    ric.R.x = state.position.x / r_mag;
    ric.R.y = state.position.y / r_mag;
    ric.R.z = state.position.z / r_mag;

    // C = h / |h| where h = r x v (angular momentum)
    Vec3 h;
    h.x = state.position.y * state.velocity.z - state.position.z * state.velocity.y;
    h.y = state.position.z * state.velocity.x - state.position.x * state.velocity.z;
    h.z = state.position.x * state.velocity.y - state.position.y * state.velocity.x;
    double h_mag = h.norm();
    ric.C.x = h.x / h_mag;
    ric.C.y = h.y / h_mag;
    ric.C.z = h.z / h_mag;

    // I = C x R (completes right-handed system)
    ric.I.x = ric.C.y * ric.R.z - ric.C.z * ric.R.y;
    ric.I.y = ric.C.z * ric.R.x - ric.C.x * ric.R.z;
    ric.I.z = ric.C.x * ric.R.y - ric.C.y * ric.R.x;

    return ric;
}

Vec3 RICFrameUtils::compute_relative_position(const StateVector& chase, const StateVector& target) {
    // Vector from target to chase in ECI
    Vec3 rel_eci;
    rel_eci.x = chase.position.x - target.position.x;
    rel_eci.y = chase.position.y - target.position.y;
    rel_eci.z = chase.position.z - target.position.z;

    // Transform to target's RIC frame
    return eci_to_ric(rel_eci, target);
}

Vec3 RICFrameUtils::compute_relative_velocity(const StateVector& chase, const StateVector& target, double n) {
    // Relative velocity in ECI
    Vec3 rel_vel_eci;
    rel_vel_eci.x = chase.velocity.x - target.velocity.x;
    rel_vel_eci.y = chase.velocity.y - target.velocity.y;
    rel_vel_eci.z = chase.velocity.z - target.velocity.z;

    // Transform to RIC (inertial measurement)
    Vec3 v_inertial = eci_to_ric(rel_vel_eci, target);

    // Get relative position for frame rotation correction
    Vec3 r_ric = compute_relative_position(chase, target);

    // In rotating frame: v_rotating = v_inertial - omega x r
    // omega = (0, 0, n) in RIC, so omega x r = (-n*I, n*R, 0)
    Vec3 v_rotating;
    v_rotating.x = v_inertial.x + n * r_ric.y;  // - (-n*I)
    v_rotating.y = v_inertial.y - n * r_ric.x;  // - (n*R)
    v_rotating.z = v_inertial.z;

    return v_rotating;
}

Vec3 RICFrameUtils::ric_to_eci(const Vec3& vec_ric, const StateVector& reference) {
    RICFrame ric = compute_frame(reference);

    Vec3 vec_eci;
    vec_eci.x = vec_ric.x * ric.R.x + vec_ric.y * ric.I.x + vec_ric.z * ric.C.x;
    vec_eci.y = vec_ric.x * ric.R.y + vec_ric.y * ric.I.y + vec_ric.z * ric.C.y;
    vec_eci.z = vec_ric.x * ric.R.z + vec_ric.y * ric.I.z + vec_ric.z * ric.C.z;

    return vec_eci;
}

Vec3 RICFrameUtils::eci_to_ric(const Vec3& vec_eci, const StateVector& reference) {
    RICFrame ric = compute_frame(reference);

    Vec3 vec_ric;
    vec_ric.x = vec_eci.x * ric.R.x + vec_eci.y * ric.R.y + vec_eci.z * ric.R.z;
    vec_ric.y = vec_eci.x * ric.I.x + vec_eci.y * ric.I.y + vec_eci.z * ric.I.z;
    vec_ric.z = vec_eci.x * ric.C.x + vec_eci.y * ric.C.y + vec_eci.z * ric.C.z;

    return vec_ric;
}

double RICFrameUtils::compute_range(const StateVector& chase, const StateVector& target) {
    Vec3 rel;
    rel.x = chase.position.x - target.position.x;
    rel.y = chase.position.y - target.position.y;
    rel.z = chase.position.z - target.position.z;
    return rel.norm();
}

double RICFrameUtils::compute_range_rate(const StateVector& chase, const StateVector& target) {
    Vec3 rel_pos;
    rel_pos.x = chase.position.x - target.position.x;
    rel_pos.y = chase.position.y - target.position.y;
    rel_pos.z = chase.position.z - target.position.z;

    Vec3 rel_vel;
    rel_vel.x = chase.velocity.x - target.velocity.x;
    rel_vel.y = chase.velocity.y - target.velocity.y;
    rel_vel.z = chase.velocity.z - target.velocity.z;

    double range = rel_pos.norm();
    if (range < 1e-10) return 0.0;

    // Range rate = (r . v) / |r|
    double dot = rel_pos.x * rel_vel.x + rel_pos.y * rel_vel.y + rel_pos.z * rel_vel.z;
    return dot / range;
}

} // namespace sim

File: visualization/cesium/js/ric_frame.js
/**
 * RIC Frame Utilities
 *
 * Port of src/targeting/ric_frame.cpp to JavaScript
 * RIC = Radial-Intrack-Crosstrack coordinate frame
 */

class RICFrame {
    /**
     * Compute RIC frame unit vectors from state vector
     * @param {Object} state - {pos: [x,y,z], vel: [x,y,z]} in ECI
     * @returns {Object} {R: [x,y,z], I: [x,y,z], C: [x,y,z]} unit vectors
     */
    static computeFrame(state) {
        const pos = state.pos;
        const vel = state.vel;

        // R = position / |position| (radial outward)
        const r_mag = Math.sqrt(pos[0]*pos[0] + pos[1]*pos[1] + pos[2]*pos[2]);
        const R = [pos[0]/r_mag, pos[1]/r_mag, pos[2]/r_mag];

        // h = r x v (angular momentum)
        const h = [
            pos[1]*vel[2] - pos[2]*vel[1],
            pos[2]*vel[0] - pos[0]*vel[2],
            pos[0]*vel[1] - pos[1]*vel[0]
        ];
        const h_mag = Math.sqrt(h[0]*h[0] + h[1]*h[1] + h[2]*h[2]);

        // C = h / |h| (cross-track, angular momentum direction)
        const C = [h[0]/h_mag, h[1]/h_mag, h[2]/h_mag];

        // I = C x R (in-track, completes right-handed system)
        const I = [
            C[1]*R[2] - C[2]*R[1],
            C[2]*R[0] - C[0]*R[2],
            C[0]*R[1] - C[1]*R[0]
        ];

        return { R, I, C };
    }

    /**
     * Compute relative position of chase w.r.t. target in RIC frame
     * @param {Object} chase - Chase state {pos: [x,y,z], vel: [x,y,z]}
     * @param {Object} target - Target state {pos: [x,y,z], vel: [x,y,z]}
     * @returns {Object} {R: meters, I: meters, C: meters}
     */
    static computeRelativePosition(chase, target) {
        // Vector from target to chase in ECI
        const rel_eci = [
            chase.pos[0] - target.pos[0],
            chase.pos[1] - target.pos[1],
            chase.pos[2] - target.pos[2]
        ];

        // Transform to target's RIC frame
        const frame = this.computeFrame(target);

        return {
            R: rel_eci[0]*frame.R[0] + rel_eci[1]*frame.R[1] + rel_eci[2]*frame.R[2],
            I: rel_eci[0]*frame.I[0] + rel_eci[1]*frame.I[1] + rel_eci[2]*frame.I[2],
            C: rel_eci[0]*frame.C[0] + rel_eci[1]*frame.C[1] + rel_eci[2]*frame.C[2]
        };
    }

    /**
     * Compute relative velocity in rotating RIC frame
     * @param {Object} chase - Chase state
     * @param {Object} target - Target state
     * @param {number} n - Mean motion (rad/s)
     * @returns {Object} {R: m/s, I: m/s, C: m/s}
     */
    static computeRelativeVelocity(chase, target, n) {
        // Relative velocity in ECI
        const rel_vel_eci = [
            chase.vel[0] - target.vel[0],
            chase.vel[1] - target.vel[1],
            chase.vel[2] - target.vel[2]
        ];

        const frame = this.computeFrame(target);

        // Project into RIC (inertial measurement)
        const v_inertial = {
            R: rel_vel_eci[0]*frame.R[0] + rel_vel_eci[1]*frame.R[1] + rel_vel_eci[2]*frame.R[2],
            I: rel_vel_eci[0]*frame.I[0] + rel_vel_eci[1]*frame.I[1] + rel_vel_eci[2]*frame.I[2],
            C: rel_vel_eci[0]*frame.C[0] + rel_vel_eci[1]*frame.C[1] + rel_vel_eci[2]*frame.C[2]
        };

        // Get relative position for frame rotation correction
        const r_ric = this.computeRelativePosition(chase, target);

        // In rotating frame: v_rotating = v_inertial - omega x r
        // omega = (0, 0, n) in RIC, so omega x r = (-n*I, n*R, 0)
        return {
            R: v_inertial.R + n * r_ric.I,
            I: v_inertial.I - n * r_ric.R,
            C: v_inertial.C
        };
    }

    /**
     * Convert RIC vector to ECI
     * @param {Object} vec_ric - {R: number, I: number, C: number}
     * @param {Object} reference - Reference state for frame definition
     * @returns {Array} [x, y, z] in ECI
     */
    static ricToECI(vec_ric, reference) {
        const frame = this.computeFrame(reference);

        return [
            vec_ric.R*frame.R[0] + vec_ric.I*frame.I[0] + vec_ric.C*frame.C[0],
            vec_ric.R*frame.R[1] + vec_ric.I*frame.I[1] + vec_ric.C*frame.C[1],
            vec_ric.R*frame.R[2] + vec_ric.I*frame.I[2] + vec_ric.C*frame.C[2]
        ];
    }

    /**
     * Convert ECI vector to RIC
     * @param {Array} vec_eci - [x, y, z] in ECI
     * @param {Object} reference - Reference state for frame definition
     * @returns {Object} {R: number, I: number, C: number}
     */
    static eciToRIC(vec_eci, reference) {
        const frame = this.computeFrame(reference);

        return {
            R: vec_eci[0]*frame.R[0] + vec_eci[1]*frame.R[1] + vec_eci[2]*frame.R[2],
            I: vec_eci[0]*frame.I[0] + vec_eci[1]*frame.I[1] + vec_eci[2]*frame.I[2],
            C: vec_eci[0]*frame.C[0] + vec_eci[1]*frame.C[1] + vec_eci[2]*frame.C[2]
        };
    }

    /**
     * Compute range between two states
     * @param {Object} chase - Chase state
     * @param {Object} target - Target state
     * @returns {number} Range in meters
     */
    static computeRange(chase, target) {
        const dx = chase.pos[0] - target.pos[0];
        const dy = chase.pos[1] - target.pos[1];
        const dz = chase.pos[2] - target.pos[2];
        return Math.sqrt(dx*dx + dy*dy + dz*dz);
    }

    /**
     * Compute range rate between two states
     * @param {Object} chase - Chase state
     * @param {Object} target - Target state
     * @returns {number} Range rate in m/s (positive = opening)
     */
    static computeRangeRate(chase, target) {
        const rel_pos = [
            chase.pos[0] - target.pos[0],
            chase.pos[1] - target.pos[1],
            chase.pos[2] - target.pos[2]
        ];

        const rel_vel = [
            chase.vel[0] - target.vel[0],
            chase.vel[1] - target.vel[1],
            chase.vel[2] - target.vel[2]
        ];

        const range = Math.sqrt(rel_pos[0]*rel_pos[0] + rel_pos[1]*rel_pos[1] + rel_pos[2]*rel_pos[2]);
        if (range < 1e-10) return 0.0;

        const dot = rel_pos[0]*rel_vel[0] + rel_pos[1]*rel_vel[1] + rel_pos[2]*rel_vel[2];
        return dot / range;
    }
}

// Export for module use
if (typeof module !== 'undefined' && module.exports) {
    module.exports = RICFrame;
}
